### ISSUE TITLE
Joins to a TransactionalIndexedCollection do not release read locks

### ISSUE THREAD

**Issue by m90smith** [npgall/cqengine#89]
I'm using CQEngine 2.7.1. I have two TransactionalIndexedCollection and when I repeatedly query one of the collections that uses the "existsIn" operator to the other collection I eventually get the following error. It appears like the read to the second collection doesn't get closed. 

Is this a bug or is there a way for me to close that ResultSet?

Here is the error I'm getting:

`java.lang.Error: Maximum lock count exceeded
	at java.util.concurrent.locks.ReentrantReadWriteLock$Sync.tryReadLock(ReentrantReadWriteLock.java:585)
	at java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.tryLock(ReentrantReadWriteLock.java:799)
	at com.googlecode.cqengine.TransactionalIndexedCollection.acquireReadLockForCurrentVersion(TransactionalIndexedCollection.java:173)
	at com.googlecode.cqengine.TransactionalIndexedCollection.retrieve(TransactionalIndexedCollection.java:366)
	at com.googlecode.cqengine.TransactionalIndexedCollection.retrieve(TransactionalIndexedCollection.java:354)
	at com.googlecode.cqengine.query.simple.ExistsIn.matchesNonSimpleAttribute(ExistsIn.java:74)
	at com.googlecode.cqengine.query.simple.SimpleQuery.matches(SimpleQuery.java:90)
	at com.googlecode.cqengine.index.fallback.FallbackIndex$1.matches(FallbackIndex.java:127)
	at com.googlecode.cqengine.resultset.common.WrappedResultSet.matches(WrappedResultSet.java:49)
	at com.googlecode.cqengine.resultset.connective.ResultSetIntersection$2.isValid(ResultSetIntersection.java:95)
	at com.googlecode.cqengine.resultset.filter.FilteringIterator.hasNext(FilteringIterator.java:52)
	at com.googlecode.cqengine.resultset.filter.FilteringIterator.hasNext(FilteringIterator.java:50)
	at com.googlecode.cqengine.resultset.filter.MaterializedDeduplicatedIterator.hasNext(MaterializedDeduplicatedIterator.java:50)
	at com.googlecode.cqengine.resultset.iterator.IteratorUtil.countElements(IteratorUtil.java:45)
	at com.googlecode.cqengine.resultset.order.MaterializedDeduplicatedResultSet.size(MaterializedDeduplicatedResultSet.java:62)
	at com.googlecode.cqengine.resultset.closeable.CloseableResultSet.size(CloseableResultSet.java:62)
	at com.googlecode.cqengine.resultset.closeable.CloseableResultSet.size(CloseableResultSet.java:62)
	at com.googlecode.cqengine.resultset.closeable.CloseableResultSet.size(CloseableResultSet.java:62)
`

**Comment by npgall** [comment]
This might be a bug. It sounds suspicious indeed!

Could you provide a self contained code snippet I can run (main method, or
unit test) which reproduces the problem?

**Comment by m90smith** [comment]
Here is a Unit Test that produces the error:

```
 public void testExistsInTransaction()
	{
		IndexedCollection colOne = new TransactionalIndexedCollection(Car.class);
		IndexedCollection colTwo = new TransactionalIndexedCollection(Car.class);
		colOne.add(new Car(1));
		colTwo.add(new Car(1));

		Query<Car> query = QueryFactory.existsIn(colTwo, Car.CAR_ID, Car.CAR_ID);

		for (int i = 0; i < 100000; i++)
		{
			ResultSet results = colOne.retrieve(query);
			results.size();
			results.close();
		}
	}

	public static class Car
	{
		public final int carId;

		public Car(int carId)
		{
			this.carId = carId;
		}

		// -------------------------- Attributes --------------------------
		public static final Attribute<Car, Integer> CAR_ID = new SimpleAttribute<Car, Integer>("carId")
		{
			public Integer getValue(Car car, QueryOptions queryOptions)
			{
				return car.carId;
			}
		};
	}
```

**Comment by vlastimil-dolejs** [comment]
Today I have came across the same problem. Bug is inside `ExistsIn` class - all calls to `foreignCollection.retrieve(..)` do not close the retrieved `ResultSet`.

In my case it will cause deadlock in `collection.update(..)` which is called after `existsIn` query.

**Comment by npgall** [comment]
Thanks for the test case!

I'll take a look at this in the next few days. Reproducing the problem is
usually half the battle, so thanks for the test case - hopefully it will be
easy to fix!

**Comment by npgall** [comment]
This is now fixed and will be included in the next release. Thanks for reporting it!

**Comment by m90smith** [comment]
I saw that this was released in 2.8.0. Has that version been pushed to Maven yet? I'm not seeing it.

**Comment by npgall** [comment]
Oh I just noticed it was still in maven staging. I've pushed it now. Thanks!
